src/goto-symex/goto_state.cpp

```diff
@@ -34,3 +36,47 @@ std::size_t goto_statet::increase_generation(
 
   return current_emplace_res.first->second.second;
 }
+
+void goto_statet::apply_condition(
+  const exprt &condition,
+  const goto_symex_statet &previous_state,
+  const namespacet &ns)
+{
+  if(condition.id() == ID_and)
+  {
+    for(const auto &op : condition.operands())
+      apply_condition(op, previous_state, ns);
+  }
+  else if(condition.id() == ID_equal)
+  {
+    exprt lhs = condition.op0();
+    exprt rhs = condition.op1();
+    if(is_ssa_expr(rhs))
+      std::swap(lhs, rhs);
+
+    if(is_ssa_expr(lhs) && goto_symex_is_constantt()(rhs))
+    {
+      const ssa_exprt &ssa_lhs = to_ssa_expr(lhs);
+      INVARIANT(
+        !ssa_lhs.get_level_2().empty(),
+        "apply_condition operand should be L2 renamed");
+
+      if(
+        previous_state.threads.size() > 1 &&
+        previous_state.write_is_shared(ssa_lhs, ns) ==
+        goto_symex_statet::write_is_shared_resultt::SHARED)
```

Comment:
Not a proper review just yet, only poking at what I guessed would be key bits:
1) Why can't the goal of this PR be addressed completely by putting the RHS value in the propagation map? (And really just this, nothing else.)
2) I would have assumed the one case where we _cannot_ do this is when we're dealing with shared variables. But now the above condition suggests that it's exactly that case that we _do_ put in the propagation map?

Reply:
1. According to our current merge logic, equal L2 generation numbers --> merge is a no-op. We must either introduce the possibility that despite equal L2 generation numbers the constant prop might disagree (moderately costly), or allocate an "administrative" L2 generation, as we do for constant assignment.
2. I accidentally flipped the condition while refactoring, fixed now.

Reply:
But doesn't this then introduce spurious merge assignments? And is it actually a problem if constant propagation disagrees?

Reply:
(the value-set assignment... just seems like we might as well; this way @owen-jones-diffblue's value-set-filtering will have less work to do as we have established a straightforward constant.

Reply:
Extra merge assignments: yes, harmless but some noise, indeed. We must do *something* about the constant propagator, or else the knowledge will outlive the code dominated by the branch or test.

Basically would you rather handle this through the L2 mechanism, or pay cost prop. to the size of the propagator map at each branch to merge them?

Reply:
Maybe it's worth adding an extra set of `irep_idt` that lists the constants inferred that way, and is then consulted at join points?

Reply:
Happy to do that -- will appeal to @kroening for an opinion